README.md
pamqp
=====

pamqp is a pure-python AMQP 0-9-1 frame encoder and decoder. The aim is to create a
client agnostic python encoder and decoder for general purpose use.

[![PyPI version](https://badge.fury.io/py/helper.png)](http://badge.fury.io/py/helper) [![Downloads](https://pypip.in/d/pamqp/badge.png)](https://crate.io/packages/pamqp) [![Build Status](https://travis-ci.org/gmr/pamqp.png?branch=master)](https://travis-ci.org/gmr/paqmp)


Documentation
-------------
http://pamqp.readthedocs.org

License
-------
BSD

Python Versions Supported
-------------------------
2.6, 2.7, 3.2, 3.3

Requirements
------------
codegen (not required for use) requires lxml.


docs/history.rst
Version History
===============

3.0.0a5 (2020-03-11)
--------------------
- Rename `pamqp.frame._frame_parts` to `pamqp.frame.frame_parts` (#15 again)

3.0.0a4 (2020-01-01)
--------------------
- Refactor codegen.py
- Revert the behaviors added in 3.0.0a2 with regard to documented defaults and `None`
- Use `amqp0-9-1.extended.xml` instead of `amqp-0-9-1.xml` to get the documentation for RabbitMQ added classes/methods
- Add strict value checking for deprecated values
- Remove empty ``__init__`` functions from method classes

3.0.0a3 (2019-12-31)
--------------------
- Make comparison of Basic.Properties against other object types raise `NotImplementedError`
- Return test coverage to 100%

3.0.0a2 (2019-12-31)
--------------------
- Added mypy as part of the test pipeline and made updates based upon its findings.
- Added length checking and regex checking for values specified in AMQP spec
- Fixed some of the type annotations added in 3.0.0a0
- Fixed some of the documentation and label usage in `pamqp.commands`
- Removed redundant inline documentation in `pamqp.commands`
- Updated default values to only reflect defaults specified in the XML and JSON specs. If no default is specified, the value will now be `None`.

3.0.0a1 (2019-12-19)
--------------------
- Revert the removal of `pamqp.body.ContentBody.name`

3.0.0a0 (2019-12-16)
--------------------
- Update to support Python 3.6+ only
- Add typing annotations to all modules, callables, and classes
- Moved exceptions from `pamqp.specification` to `pamqp.exceptions`
- Moved constants from `pamqp.specification` to `pamqp.constants`
- Moved base classes out of `pamqp.specification` to `pamqp.base`
- Changed the structure of nested classes for AMQP Commands (Classes & Methods) in `pamqp.specification` to functions in `pamqp.commands`
- Renamed `pamqp.specification.ERRORS` to `pamqp.exceptions.CLASS_MAPPING`
- Remove convenience exports of `pamqp.headers.ContentHeader` and `pamqp.header.ProtocolHeader`
- pamqp.body.ContentBody.value now only supports `bytes`
- Changed `pamqp.decode.timestamp` to return a `datetime.datetime` instance instead of `time.struct_time`.
- Updated `pamqp.encode.support_deprecated_rabbitmq()` to allow for toggling support.
- Changed `pamqp.encode.timestamp` to only support `datetime.datetime` and `time.struct_time` values, dropping epoch (`int`) support.
- Removed `pamqp.frame.BasicProperties.to_dict()` in favor of behavior allowing for `dict(pamqp.frame.BasicProperties)`
- Optimized `pamqp.heartbeat.Heartbeat` to marshal the static frame value as a predefined class attribute.
- Add support for `Connection.UpdateSecret` and `Connection.UpdateSecretOk`.
- Removed the ability to unset a `Basic.Property` by invoking `del properties[key]`
- Removed the deprecated `pamqp.codec` sub-package

2.3.0 (2019-04-18)
------------------
- Add :py:func:`pamqp.encode.support_deprecated_rabbitmq` function to limit data types available when encoding field-tables for older RabbitMQ versions.

2.2.0 (2019-04-18)
------------------
- Change :py:meth:`pamqp.encode.timestamp` to allow for numeric/epoch timestamps (#14 - `mosquito <https://github.com/mosquito>`_)
- Change :py:meth:`pamqp.frame.frame_parts` to a public method (#15 - `mosquito <https://github.com/mosquito>`_)
- Cleanup of code to pass configured flake8 tests
- Add support for 8-bit unsigned integer values in :py:meth:`pamqp.encode.table_integer`

2.1.0 (2018-12-28)
------------------
- Change raising a DeprecationWarning exception to using warnings.warn for deprecated AMQP methods (#13 - `dzen <https://github.com/dzen>`_)

2.0.0 (2018-09-11)
------------------
- **Change Python versions supported to 2.7 and 3.4+**
- **Always decode field table keys as strings (#6)**
   - This may be a breaking change means in Python3 keys will always be type str for short strings. This includes frame
     values and field table values.
   - In Python 2.7 if a short-string (key, frame field value, etc) has UTF-8 characters in it, it will be a `unicode` object.
- Combine test coverage across all Python versions
- Fix range for signed short integer (#7)
- Fix guards for usage of unsigned short usage in `pamqp.encode` (#7)
- Fix encoding and decoding of unsigned short (#7)
- Add support for unsigned short integer and long integer in field tables  (#10)
- Address edge case of small value in long type (#8)
- Address long string encoding inconsistency (#9)
- Cleanup unicode object & conditionals in py3 (#9)
- Add `pamqp.exceptions.PAMQPException` as a base class for pamqp specific exceptions (#4)
- Fix decoding of void values in a field table or array

1.6.1 (2015-02-05)
------------------
- Fix the encoding guard for unsigned short integers to be 65535 [rabbitpy #62]

1.6.0 (2014-12-12)
------------------
- Remove UTF-8 encoding from byte_array (#2)
- Fix AMQP Field Tables / `Basic.Properties` headers behavior:
   - Field names per spec should not exceed 128 bytes
   - long-strings should not be utf-8 encoded (only short-strings *boggle*)
- Ensure that field table long strings are not coerced to UTF-8 as specified in AMQP 0-9-1
   If a string is passed in as a long string in a field table and it contains UTF-8 characters it will be UTF-8 encoded
- Move AMQP Methods in specification.py to slotted classes
- Change `Basic.Properties` to a slotted class
- Instead of class level attributes with the same name as obj attributes, prefix class attributes for data types with an underscore
- Add new class method type() for `Basic.Properties` for accessing data type
- Add new class method type() for AMQP methods for accessing data type
- Change `Basic.Properties.attributes` to `Basic.Properties.attributes()`, returning the list of slotted attributes
- Fix a typo for booleans in the method mapping for table decoding
- `Frame.__getitem__` will now raise a KeyError instead of None for an invalid attribute
- `PropertiesBase` no longer checks to see if an attribute is set for contains
- Adds new specification tests

pamqp/encode.py
# -*- encoding: utf-8 -*-
"""
Functions for encoding data of various types including field tables and arrays

"""
import calendar
import datetime
import decimal as _decimal
import logging
import struct
import time
import typing

from pamqp import common

LOGGER = logging.getLogger(__name__)

DEPRECATED_RABBITMQ_SUPPORT = False
"""Toggle to support older versions of RabbitMQ."""


def support_deprecated_rabbitmq(enabled: bool = True) -> None:
    """Toggle the data types available in field-tables

    If called with `True`, than RabbitMQ versions, the field-table integer
    types will not support the full AMQP spec.

    :param enabled: Specify if deprecated RabbitMQ versions are supported

    """
    global DEPRECATED_RABBITMQ_SUPPORT

    DEPRECATED_RABBITMQ_SUPPORT = enabled


def by_type(value: common.FieldValue, data_type: str) -> bytes:
    """Takes a value of any type and tries to encode it with the specified
    encoder.

    :param value: The value to encode
    :type value: :const:`pamqp.common.FieldValue`
    :param data_type: The data type name to use for encoding
    :raises TypeError: when the :data:`data_type` is unknown

    """
    try:
        return METHODS[str(data_type)](value)
    except KeyError:
        raise TypeError('Unknown type: {}'.format(value))


def bit(value: int, byte: int, position: int) -> int:
    """Encode a bit value

    :param value: Value to encode
    :param byte: The byte to apply the value to
    :param position: The position in the byte to set the bit on

    """
    return byte | (value << position)


def boolean(value: bool) -> bytes:
    """Encode a boolean value

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type

    """
    if not isinstance(value, bool):
        raise TypeError('bool required, received {}'.format(type(value)))
    return common.Struct.short_short.pack(int(value))


def byte_array(value: bytearray) -> bytes:
    """Encode a byte array value

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type

    """
    if not isinstance(value, bytearray):
        raise TypeError('bytearray required, received {}'.format(type(value)))
    return common.Struct.integer.pack(len(value)) + value


def decimal(value: _decimal.Decimal) -> bytes:
    """Encode a decimal.Decimal value

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type

    """
    if not isinstance(value, _decimal.Decimal):
        raise TypeError('decimal.Decimal required, received {}'.format(
            type(value)))
    tmp = str(value)
    if '.' in tmp:
        decimals = len(tmp.split('.')[-1])
        value = value.normalize()
        raw = int(value * (_decimal.Decimal(10)**decimals))
        return struct.pack('>Bi', decimals, raw)
    return struct.pack('>Bi', 0, int(value))


def double(value: float) -> bytes:
    """Encode a floating point value as a double

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type

    """
    if not isinstance(value, float):
        raise TypeError('float required, received {}'.format(type(value)))
    return common.Struct.double.pack(value)


def floating_point(value: float) -> bytes:
    """Encode a floating point value

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type

    """
    if not isinstance(value, float):
        raise TypeError('float required, received {}'.format(type(value)))
    return common.Struct.float.pack(value)


def long_int(value: int) -> bytes:
    """Encode a long integer

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type or outside the
        acceptable range for the data type

    """
    if not isinstance(value, int):
        raise TypeError('int required, received {}'.format(type(value)))
    elif not (-2147483648 <= value <= 2147483647):
        raise TypeError('Long integer range: -2147483648 to 2147483647')
    return common.Struct.long.pack(value)


def long_uint(value: int) -> bytes:
    """Encode a long unsigned integer

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type or outside the
        acceptable range for the data type

    """
    if not isinstance(value, int):
        raise TypeError('int required, received {}'.format(type(value)))
    elif not (0 <= value <= 4294967295):
        raise TypeError('Long unsigned-integer range: 0 to 4294967295')
    return common.Struct.ulong.pack(value)


def long_long_int(value: int) -> bytes:
    """Encode a long-long int

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type or outside the
        acceptable range for the data type

    """
    if not isinstance(value, int):
        raise TypeError('int required, received {}'.format(type(value)))
    elif not (-9223372036854775808 <= value <= 9223372036854775807):
        raise TypeError('long-long integer range: '
                        '-9223372036854775808 to 9223372036854775807')
    return common.Struct.long_long_int.pack(value)


def long_string(value: str) -> bytes:
    """Encode a "long string"

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type

    """
    return _string(common.Struct.integer, value)


def octet(value: int) -> bytes:
    """Encode an octet value

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type

    """
    if not isinstance(value, int):
        raise TypeError('int required, received {}'.format(type(value)))
    return common.Struct.byte.pack(value)


def short_int(value: int) -> bytes:
    """Encode a short integer

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type or outside the
        acceptable range for the data type

    """
    if not isinstance(value, int):
        raise TypeError('int required, received {}'.format(type(value)))
    elif not (-32768 <= value <= 32767):
        raise TypeError('Short integer range: -32678 to 32767')
    return common.Struct.short.pack(value)


def short_uint(value: int) -> bytes:
    """Encode an unsigned short integer

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type or outside the
        acceptable range for the data type

    """
    if not isinstance(value, int):
        raise TypeError('int required, received {}'.format(type(value)))
    elif not (0 <= value <= 65535):
        raise TypeError('Short unsigned integer range: 0 to 65535')
    return common.Struct.ushort.pack(value)


def short_string(value: str) -> bytes:
    """ Encode a string

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type

    """
    return _string(common.Struct.byte, value)


def timestamp(value: typing.Union[datetime.datetime, time.struct_time]) \
        -> bytes:
    """Encode a datetime.datetime object or time.struct_time

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type

    """
    if isinstance(value, datetime.datetime):
        value = value.timetuple()
    if isinstance(value, time.struct_time):
        return common.Struct.timestamp.pack(calendar.timegm(value))
    raise TypeError(
        'datetime.datetime or time.struct_time required, received {}'.format(
            type(value)))


def field_array(value: common.FieldArray) -> bytes:
    """Encode a field array from a list of values

    :param value: Value to encode
    :type value: :const:`pamqp.common.FieldArray`
    :raises TypeError: when the value is not the correct type

    """
    if not isinstance(value, list):
        raise TypeError('list of values required, received {}'.format(
            type(value)))
    data = []
    for item in value:
        data.append(encode_table_value(item))
    output = b''.join(data)
    return common.Struct.integer.pack(len(output)) + output


def field_table(value: common.FieldTable) -> bytes:
    """Encode a field table from a dict

    :param value: Value to encode
    :type value: :const:`pamqp.common.FieldTable`
    :raises TypeError: when the value is not the correct type

    """
    if not value:  # If there is no value, return a standard 4 null bytes
        return common.Struct.integer.pack(0)
    elif not isinstance(value, dict):
        raise TypeError('dict required, received {}'.format(type(value)))
    data = []
    for key, value in sorted(value.items()):
        if len(key) > 128:  # field names have 128 char max
            LOGGER.warning('Truncating key %s to 128 bytes', key)
            key = key[0:128]
        data.append(short_string(key))
        try:
            data.append(encode_table_value(value))
        except TypeError as err:
            raise TypeError('{} error: {}/'.format(key, err))
    output = b''.join(data)
    return common.Struct.integer.pack(len(output)) + output


def table_integer(value: int) -> bytes:
    """Determines the best type of numeric type to encode value as, preferring
    the smallest data size first.

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type or outside the
        acceptable range for the data type

    """
    if DEPRECATED_RABBITMQ_SUPPORT:
        return _deprecated_table_integer(value)
    if 0 <= value <= 255:
        return b'b' + octet(value)
    elif -32768 <= value <= 32767:
        return b's' + short_int(value)
    elif 0 <= value <= 65535:
        return b'u' + short_uint(value)
    elif -2147483648 <= value <= 2147483647:
        return b'I' + long_int(value)
    elif 0 <= value <= 4294967295:
        return b'i' + long_uint(value)
    elif -9223372036854775808 <= value <= 9223372036854775807:
        return b'l' + long_long_int(value)
    raise TypeError('Unsupported numeric value: {}'.format(value))


def _deprecated_table_integer(value: int) -> bytes:
    """Determines the best type of numeric type to encode value as, preferring
    the smallest data size first, supporting versions of RabbitMQ < 3.6

    :param value: Value to encode
    :raises TypeError: when the value is not the correct type or outside the
        acceptable range for the data type

    """
    if 0 <= value <= 255:
        return b'b' + octet(value)
    elif -32768 <= value <= 32767:
        return b's' + short_int(value)
    elif -2147483648 <= value <= 2147483647:
        return b'I' + long_int(value)
    elif -9223372036854775808 <= value <= 9223372036854775807:
        return b'l' + long_long_int(value)
    raise TypeError('Unsupported numeric value: {}'.format(value))


def _string(encoder: struct.Struct, value: str) -> bytes:
    """Reduce a small amount of duplication in string handling

    :raises: TypeError

    """
    if not isinstance(value, str):
        raise TypeError('str required, received {}'.format(type(value)))
    temp = value.encode('utf-8')
    return encoder.pack(len(temp)) + temp


def encode_table_value(value: typing.Union[common.FieldArray,
                                           common.FieldTable,
                                           common.FieldValue]) -> bytes:
    """Takes a value of any type and tries to encode it with the proper encoder

    :param value: Value to encode
    :type value: :const:`pamqp.common.FieldArray` or
                 :const:`pamqp.common.FieldTable` or
                 :const:`pamqp.common.FieldValue`
    :raises TypeError: when the type of the value is not supported

    """
    if isinstance(value, bool):
        return b't' + boolean(value)
    elif isinstance(value, int):
        return table_integer(value)
    elif isinstance(value, _decimal.Decimal):
        return b'D' + decimal(value)
    elif isinstance(value, float):
        return b'f' + floating_point(value)
    elif isinstance(value, str):
        return b'S' + long_string(value)
    elif isinstance(value, (datetime.datetime, time.struct_time)):
        return b'T' + timestamp(value)
    elif isinstance(value, dict):
        return b'F' + field_table(value)
    elif isinstance(value, list):
        return b'A' + field_array(value)
    elif isinstance(value, bytearray):
        return b'x' + byte_array(value)
    elif value is None:
        return b'V'
    raise TypeError('Unknown type: {} ({!r})'.format(type(value), value))


METHODS = {
    'bytearray': byte_array,
    'double': double,
    'field_array': field_array,
    'long': long_uint,
    'longlong': long_long_int,
    'longstr': long_string,
    'octet': octet,
    'short': short_uint,
    'shortstr': short_string,
    'table': field_table,
    'timestamp': timestamp,
    'void': lambda _: None,
}
